Mathlib/LinearAlgebra/Matrix/SemiringInverse.lean

```diff
@@ -0,0 +1,275 @@
+    contrapose! hσ
+    rw [not_mem_compl, mem_map, ofSign_disjUnion]
+    exact ⟨Equiv.ofBijective σ hσ.bijective_of_finite, mem_univ _, rfl⟩
+  obtain ⟨i, j, hσ, hij⟩ := Function.not_injective_iff.mp hσ
+  replace hσ k : σ (swap i j k) = σ k := by
+    rw [swap_apply_def]
+    split_ifs with h h <;> simp only [hσ, h]
+  rw [← mul_neg_one, hf (mem_ofSign.mpr (sign_swap hij)), sum_map]
+  simp_rw [prod_mul_distrib, mulRightEmbedding_apply, Perm.mul_apply]
+  refine sum_congr rfl fun τ hτ ↦ congr_arg (_ *  ·) ?_
+  rw [← Equiv.prod_comp (swap i j)]
+  simp only [hσ]
+
+theorem mul_adjp_apply_eq : (A * adjp s A) i i = detp s A := by
+  have key := sum_fiberwise_eq_sum_filter (ofSign s) univ (· i) fun σ ↦ ∏ k, A k (σ k)
+  simp_rw [mem_univ, filter_True] at key
+  simp_rw [mul_apply, adjp_apply, mul_sum, detp, ← key]
+  refine sum_congr rfl fun x hx ↦ sum_congr rfl fun σ hσ ↦ ?_
+  rw [← prod_mul_prod_compl ({i} : Finset n), prod_singleton, (mem_filter.mp hσ).2]
+
+theorem mul_adjp_apply_ne (h : i ≠ j) : (A * adjp 1 A) i j = (A * adjp (-1) A) i j := by
+  simp_rw [mul_apply, adjp_apply, mul_sum, sum_sigma']
+  let f : (Σ x : n, Perm n) → (Σ x : n, Perm n) := fun ⟨x, σ⟩ ↦ ⟨σ i, σ * swap i j⟩
+  let t s : Finset (Σ x : n, Perm n) := univ.sigma fun x ↦ (ofSign s).filter fun σ ↦ σ j = x
+  have hf {s} : ∀ p ∈ t s, f (f p) = p := by
+    intro ⟨x, σ⟩ hp
+    rw [mem_sigma, mem_filter, mem_ofSign] at hp
+    simp_rw [f, Perm.mul_apply, swap_apply_left, hp.2.2, mul_swap_mul_self]
+  refine sum_bij' (fun p _ ↦ f p) (fun p _ ↦ f p) ?_ ?_ hf hf ?_
+  · intro ⟨x, σ⟩ hp
+    rw [mem_sigma, mem_filter, mem_ofSign] at hp ⊢
+    rw [Perm.mul_apply, sign_mul, hp.2.1, sign_swap h, swap_apply_right]
+    exact ⟨mem_univ (σ i), rfl, rfl⟩
+  · intro ⟨x, σ⟩ hp
+    rw [mem_sigma, mem_filter, mem_ofSign] at hp ⊢
+    rw [Perm.mul_apply, sign_mul, hp.2.1, sign_swap h, swap_apply_right]
+    exact ⟨mem_univ (σ i), rfl, rfl⟩
+  · intro ⟨x, σ⟩ hp
+    rw [mem_sigma, mem_filter, mem_ofSign] at hp
+    have key : ({j}ᶜ : Finset n) = disjUnion ({i} : Finset n) ({i, j} : Finset n)ᶜ (by simp) := by
+      rw [singleton_disjUnion, cons_eq_insert, compl_insert, insert_erase]
+      rwa [mem_compl, mem_singleton]
+    simp_rw [key, prod_disjUnion, prod_singleton, Perm.mul_apply, swap_apply_left, ← mul_assoc]
+    rw [mul_comm (A i x) (A i (σ i)), hp.2.2]
+    refine congr_arg _ (prod_congr rfl fun x hx ↦ ?_)
+    rw [mem_compl, mem_insert, mem_singleton, not_or] at hx
+    rw [swap_apply_of_ne_of_ne hx.1 hx.2]
+
+theorem mul_adjp_add_detp : A * adjp 1 A + detp (-1) A • 1 = A * adjp (-1) A + detp 1 A • 1 := by
+  ext i j
+  rcases eq_or_ne i j with rfl | h <;> simp_rw [add_apply, smul_apply, smul_eq_mul]
+  · simp_rw [mul_adjp_apply_eq, one_apply_eq, mul_one, add_comm]
+  · simp_rw [mul_adjp_apply_ne A i j h, one_apply_ne h, mul_zero]
+
+variable {A B}
+
+theorem isAddUnit_mul (hAB : A * B = 1) (i j k : n) (hij : i ≠ j) : IsAddUnit (A i k * B k j) := by
```

Comment:
This lemma is also true for `A * B = scalar c`.
Which other parts of this PR generalize to that case?

Reply:
The `isAddUnit` lemmas should all generalize to diagonal matrices (although that would currently result in a circular import since the `NonsingularInverse` file is quite low-level). I think the conclusion `A * B = 1 ↔ B * A = 1` should generalize to `scalar c` with `IsUnit c`. But this generalization can be proved directly from the identity-matrix-version, which would probably be clearer than trying to shoehorn `scalar c` into the existing proofs.